### ISSUE TITLE
Metrics incorrect for RNN with mask

### ISSUE THREAD

**Issue by seandaug** [tensorflow/tensorflow#34158]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): **yes**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): **Windows 10**
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: **n/a**
- TensorFlow installed from (source or binary): **binary (conda)**
- TensorFlow version (use command below): **2.0.0**
- Python version: **3.6.8**
- Bazel version (if compiling from source): **n/a**
- GCC/Compiler version (if compiling from source): **n/a**
- CUDA/cuDNN version: **10.0.130 / 7.6.0**
- GPU model and memory: **1080 Ti**

**Describe the current behavior**
I am training an RNN (GRU) where my varying-length sequences are right-padded with 0s and a mask is applied. _Many_ sequences are more than half 0s (padding). I compile the model with a loss of `'mean_squared_error'` and a metric of `'mean_squared_error'`, but the output is different when the mask is in effect.
```
model.compile(optimizer=keras.optimizers.RMSprop(),
              loss='mean_squared_error',
              metrics=['mean_squared_error'])
```
Or equivalently:
```
model.compile(optimizer=keras.optimizers.RMSprop(),
              loss=keras.losses.MeanSquaredError(),
              metrics=[keras.metrics.MeanSquaredError()])
```

Example output (note the different values for `loss` vs. `mean_squared_error` for both training and validation):
```
Epoch 1/50
210328/210328 [==============================] - 610s 3ms/sample - loss: 4.5338e-06 - mean_squared_error: 1.1923e-05 - val_loss: 2.5456e-06 - val_mean_squared_error: 6.7928e-06
Epoch 2/50
210328/210328 [==============================] - 525s 2ms/sample - loss: 2.1835e-06 - mean_squared_error: 5.7421e-06 - val_loss: 2.2920e-06 - val_mean_squared_error: 6.1160e-06
Epoch 3/50
210328/210328 [==============================] - 513s 2ms/sample - loss: 1.9939e-06 - mean_squared_error: 5.2437e-06 - val_loss: 2.2535e-06 - val_mean_squared_error: 6.0133e-06
...
Epoch 50/50
210328/210328 [==============================] - 527s 3ms/sample - loss: 1.5595e-06 - mean_squared_error: 4.1011e-06 - val_loss: 1.7867e-06 - val_mean_squared_error: 4.7677e-06
```

When I disable the masking, I get the following output:
```
Epoch 1/3
210328/210328 [==============================] - 516s 2ms/sample - loss: 7.1682e-06 - mean_squared_error: 7.1682e-06 - val_loss: 6.7091e-06 - val_mean_squared_error: 6.7091e-06
Epoch 2/3
210328/210328 [==============================] - 434s 2ms/sample - loss: 5.9133e-06 - mean_squared_error: 5.9133e-06 - val_loss: 6.7091e-06 - val_mean_squared_error: 6.7091e-06
Epoch 3/3
210328/210328 [==============================] - 442s 2ms/sample - loss: 5.9085e-06 - mean_squared_error: 5.9085e-06 - val_loss: 6.7073e-06 - val_mean_squared_error: 6.7073e-06
```
Without the mask, the values for `loss` and `mean_squared_error` agree. For the validation set, the values are not really improving and the value of 6.7e-06 seems to be what you get when you evaluate on the 0s that would otherwise be ignored by the masking. Comparing the values between the runs suggests that the `mean_squared_error` calculations are not using the mask when it is in effect, but the `loss` calculations do use the mask. (We'd expect lower values when we correctly ignore irrelevant time steps.)

**Describe the expected behavior**
The values for `loss` and `mean_squared_error` should agree and both use the masking.

**Code to reproduce the issue**
I don't have full code and data to share since my model and data are proprietary.

**Other info / logs**
I can't think of any relevant logs.

**Comment by gadagashwini** [comment]
@seandaug, Please provide the complete code to reproduce the reported issue. Thanks!

**Comment by seandaug** [comment]
Here's a small example that illustrates the issue.
```
import tensorflow as tf
from tensorflow import keras
import numpy as np

# %% Make some sine wave data
num_time_series = 100
max_length = 200
num_points = 1
training_data = np.zeros((num_time_series, max_length, num_points))
np.random.seed(123)
tf.random.set_seed(123)
for i in range(num_time_series):
    # Make a sequence that doesn't fill the array so there's padding at the end
    length = np.random.randint(0.25 * max_length, .75 * max_length + 1)
    period = np.random.random() * 20 + 5
    shift = np.random.random()
    training_data[i, 0:length, 0] = np.sin(2 * np.pi / period * np.linspace(0, length - 1, length) + shift)


# %% Define the model
def make_model(use_mask):
    input_seq = keras.layers.Input(shape=(None, num_points))
    masked_input_seq = keras.layers.Masking(mask_value=0.0)(input_seq) if use_mask else input_seq
    gru = keras.layers.GRU(units=3, return_sequences=True)(masked_input_seq)
    output = keras.layers.Dense(units=1)(gru)
    model = keras.Model(input_seq, output)
    model.compile(optimizer=keras.optimizers.RMSprop(),
                  loss='mean_squared_error',
                  metrics=['mean_squared_error'])
    return model


# %% Train model with mask. Note that reported 'loss' and 'mean_squared_error' differ
make_model(True).fit(training_data[:, :-1, :],
                     training_data[:, 1:, :],
                     batch_size=10,
                     epochs=3,
                     verbose=1)


# %% Train model without mask. Note that reported 'loss' and 'mean_squared_error' match
make_model(False).fit(training_data[:, :-1, :],
                      training_data[:, 1:, :],
                      batch_size=10,
                      epochs=3,
                      verbose=1)
```

Example output when mask is used (note that `loss` and `mean_squared_error` differ):
```
Epoch 1/3
100/100 [==============================] - 10s 98ms/sample - loss: 0.2741 - mean_squared_error: 0.5393
Epoch 2/3
100/100 [==============================] - 2s 20ms/sample - loss: 0.2690 - mean_squared_error: 0.5292
Epoch 3/3
100/100 [==============================] - 2s 20ms/sample - loss: 0.2650 - mean_squared_error: 0.5214
```

Example output when no mask is used (note that `loss` and `mean_squared_error` match):
```
Epoch 1/3
100/100 [==============================] - 4s 37ms/sample - loss: 0.2475 - mean_squared_error: 0.2475
Epoch 2/3
100/100 [==============================] - 1s 13ms/sample - loss: 0.2305 - mean_squared_error: 0.2305
Epoch 3/3
100/100 [==============================] - 1s 13ms/sample - loss: 0.2174 - mean_squared_error: 0.2174
```

**Comment by gadagashwini** [comment]
I could replicate the issue with Tf 2.0. 
Please take a look at the gist [here](https://colab.sandbox.google.com/gist/gadagashwini/740c0fd036e17fe505996957b79d5cbb/untitled255.ipynb). Thanks!

**Comment by qlzh727** [comment]
Adding @pavithrasv who is the owner of metric and loss. I think the loss will take into account of the masks and exclude the masked value, but I don't think metric will do that, which is why you see the value difference here. I will let @pavithrasv  to confirm.

**Comment by jvishnuvardhan** [comment]
@seandaug I think this was resolved in the `tf-nightly`. I am not able to reproduce the error with `tf-nightly`. Please take a look at the [gist here](https://colab.sandbox.google.com/gist/jvishnuvardhan/6e844af327510d58c27d59f78593825d/untitled255.ipynb). Thanks!

I am closing this issue as this was resolved. Please feel free to reopen if I am mistaken. Thanks!

**Comment by seandaug** [comment]
@jvishnuvardhan The gist you posted using `tf-nightly` does not appear to work anymore. The issue seems to have reappeared.

**Comment by seandaug** [comment]
It appears that the difference is in how the "mean" is computed over the values for a "loss" vs. a "metric". Consider this very simple example that just outputs the input after applying a mask:
```
import tensorflow as tf
import numpy as np
y_true = np.array([[[0.], [1.], [1.]]])
x = np.array([[[1.], [0.], [1.]]])
model = tf.keras.Sequential([tf.keras.layers.Masking(mask_value=0., input_shape=(3, 1))])
model.compile(loss='mae', metrics=['mae'])
model.fit(x, y_true)  # Would intuitively expect mask to give loss of 0.5, but get 0.3333
```
This generates the following output:
```
1/1 [==============================] - 0s 432us/step - loss: 0.3333 - mae: 0.5000
```
Here we have a sequence of length 3, the second value of which is masked out. Therefore, the MAE should be 0.5 because the total absolute error is 1.0 and it is divided among 2 entries (after applying the mask). However, as a "loss", the MAE is reported as 0.3333 instead of 0.5.

I suspect this is because of the use of `reduction=losses_utils.ReductionV2.AUTO` in `tf.keras.losses.Loss` that relies on `losses_utils.ReductionV2.SUM_OVER_BATCH_SIZE`. Then the denominator of the MAE calculation is the batch size (including sequence length) that does not consider the mask. Somehow the use of MAE as a "metric" (not a "loss") avoids this issue.

We get similar intuitively unexpected results when using `sample_weight` as well (no masking involved). Consider this simple example that just outputs the input and allows sample weights to be specified:
```
import tensorflow as tf
import numpy as np
y_true = np.array([[0.], [1.]])
x = np.array([[1.], [0.]])
weights = np.array([1., 0.])
model = tf.keras.Sequential([tf.keras.layers.InputLayer(input_shape=(1,))])
model.compile(loss='mae', metrics=['mae'])
model.fit(x, y_true, sample_weight=weights)  # Might expect to see 1.0, but get 0.5
```
This generates the following output:
```
1/1 [==============================] - 0s 427us/step - loss: 0.5000 - mae: 1.0000
```
Here we have two samples, each with a loss of 1.0, but due to the provided sample weights, one of them should be ignored. Thus, we would expect to see an MAE of 1.0, which is correctly reported using the "metric". However, the "loss" does not match intuition because it is dividing by the batch size, not the total sample weight.

So to me, the big question here is: **Why do loss functions use `SUM_OVER_BATCH_SIZE` that ignores masking and sample weights?**

**Comment by Recepcan** [comment]
I agree with @seandaug. Loss and metrics differ if I enable masking on an LSTM model. I am trying to find a way to solve this problem as well.

**Comment by Taoup** [comment]
Hi. I've encountered this problem as well. Is there any update on this issue?

**Comment by seandaug** [comment]
@Recepcan @Taoup See https://github.com/tensorflow/tensorflow/issues/34491#issuecomment-869086784 for a workaround.
